### PR TITLE
Review

### PR DESCRIPTION
**Pull request by MunteanuAndrei237**

### CODE REVIEW

File: app/controllers/custom_issues_controller.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class CustomIssuesController < ApplicationController
+  skip_before_action :verify_authenticity_token, only: :update_collapsed_ids
```
Comment (mircea-negreanu) [inline app/controllers/custom_issues_controller.rb]: 1/ Why skip `verify_authentication_token`? Did something crashed when sending data? We need to correctly work in any scenario.
2/ I think `update_collaped_ids` should be be in the `issue_view_columns_controller.rb`, since it is related.

File: app/controllers/custom_issues_controller.rb [hunk 2]
```diff
@@ -0,0 +1,15 @@
+class CustomIssuesController < ApplicationController
+  skip_before_action :verify_authenticity_token, only: :update_collapsed_ids
+
+  #Controller used for modifying collapsed_ids field of an issue
+  def update_collapsed_ids
+    @issue = Issue.find(params[:id])
+
+    json_data = JSON.parse(request.body.read)
+    if @issue.update(collapsed_ids: json_data["collapsed_ids"])
+      render json: { message: "Collapsed IDs updated successfully" }, status: :ok
```
Comment (mircea-negreanu) [inline app/controllers/custom_issues_controller.rb]: Please move the texts to the locales. Just fill the english one.

File: app/controllers/custom_issues_controller.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+class CustomIssuesController < ApplicationController
+  skip_before_action :verify_authenticity_token, only: :update_collapsed_ids
+
+  #Controller used for modifying collapsed_ids field of an issue
+  def update_collapsed_ids
+    @issue = Issue.find(params[:id])
+
+    json_data = JSON.parse(request.body.read)
```
Comment (mircea-negreanu) [inline app/controllers/custom_issues_controller.rb]: Also, check the errors from parse, and return an error if an exception occurs.

File: app/controllers/issue_view_columns_controller.rb [hunk 4]
```diff
@@ -25,6 +25,6 @@ def update
       c.save!
     end
 
-    redirect_to settings_project_path(@project, tab: 'issue_view_columns'), notice: l(:label_issue_columns_created_sucessfully)
+    redirect_to settings_project_path(@project, tab: "issue_view_columns"), notice: l(:label_issue_columns_created_sucessfully)
```
Comment (mircea-negreanu) [inline app/controllers/issue_view_columns_controller.rb]: Why transforming the `'` to `"`? Is this an automated process? If so, can you disable it and revert these types of modifications?

File: app/helpers/issue_view_columns_issues_helper.rb [hunk 5]
```diff
@@ -2,53 +2,242 @@
...
+        column_name, min_width = column_setting.split(":").map(&:strip)
+        min_widths[column_name] = min_width
+      end
+    end
 
-      tr_classes = +"hascontextmenu #{child.css_classes} #{cycle 'odd', 'even'}"
-      tr_classes << " idnt idnt-#{level}" if level.positive?
+    # Retrieve sorting settings and determine if sorting by directory/file model is enabled
+    sort_dir_file_model = RedmineIssueViewColumns.setting(:sort_dir_file_model)
+    collapsed_ids = issue.collapsed_ids.to_s.split.map(&:to_i)
```
Comment (mircea-negreanu) [inline app/helpers/issue_view_columns_issues_helper.rb:21]: I'm curious, what is returned here if `collapsed_ids` is not completed (I think the default is `nil`, right?

File: db/migrate/003_add_collapsed_ids_to_issues.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class AddCollapsedIdsToIssues < ActiveRecord::Migration[6.0]
+  def change
+    add_column :issues, :collapsed_ids, :text
```
Comment (mircea-negreanu) [inline db/migrate/003_add_collapsed_ids_to_issues.rb]: Did you managed to check if `:test` is universal? (e.g Postgres/Mysql/sqlite)

File: app/helpers/issue_view_columns_issues_helper.rb [hunk 7]
```diff
@@ -2,53 +2,242 @@
 
 module IssueViewColumnsIssuesHelper
   def render_descendants_tree(issue)
-    columns_list = get_fields_for_project issue
-    # no field defined, then use render from core redmine (or whatever by other plugins loaded before this)
-    return super if columns_list.count.zero?
+    # Retrieve the list of columns to display for the project
+    columns_list = get_fields_for_project(issue)
```
Comment (mircea-negreanu) [inline app/helpers/issue_view_columns_issues_helper.rb]: I see this section of the code is the same as above. Why did you change it?

File: app/helpers/issue_view_columns_issues_helper.rb [hunk 8]
```diff
@@ -62,32 +251,32 @@ def render_issue_relations(issue, relations)
       other_issue = relation.other_issue issue
       next if other_issue.closed? && !issue_columns_with_closed_issues?
 
-      tr_classes = "hascontextmenu #{other_issue.css_classes} #{cycle 'odd', 'even'} #{relation.css_classes_for other_issue}"
+      tr_classes = "hascontextmenu #{other_issue.css_classes} #{cycle "odd", "even"} #{relation.css_classes_for other_issue}"
```
Comment (mircea-negreanu) [inline app/helpers/issue_view_columns_issues_helper.rb]: Here again the modification from `"` to `'`.

File: app/helpers/issue_view_columns_issues_helper.rb [hunk 9]
```diff
@@ -62,32 +251,32 @@ def render_issue_relations(issue, relations)
...
-                          relation_path(relation),
-                          remote: true,
-                          method: :delete,
-                          data: { confirm: l(:text_are_you_sure) },
-                          title: l(:label_relation_delete),
-                          class: 'icon-only icon-link-break'
-                else
-                  ''.html_safe
-                end
+          link_to l(:label_relation_delete),
```
Comment (mircea-negreanu) [inline app/helpers/issue_view_columns_issues_helper.rb]: Here you changed the indentation. Can you put it back? It makes it hard to distinguish just the functional modifications.

File: app/controllers/custom_issues_controller.rb [hunk 10]
```diff
@@ -0,0 +1,15 @@
+class CustomIssuesController < ApplicationController
+  skip_before_action :verify_authenticity_token, only: :update_collapsed_ids
+
+  #Controller used for modifying collapsed_ids field of an issue
+  def update_collapsed_ids
+    @issue = Issue.find(params[:id])
```
Comment (mircea-negreanu) [inline app/controllers/custom_issues_controller.rb]: I think we also need to check if the current user has rights to modify the issue.

File: db/migrate/003_add_collapsed_ids_to_issues.rb (hunk 6 above)
Comment (MunteanuAndrei237) [reply db/migrate/003_add_collapsed_ids_to_issues.rb]: its not in oracle SQL so I will switch to VARCHAR which is more universal

File: app/helpers/issue_view_columns_issues_helper.rb (hunk 5 above)
Comment (MunteanuAndrei237) [reply app/helpers/issue_view_columns_issues_helper.rb:21]: collapsed_ids will become an empty array and all the issues will be rendered normally

File: app/controllers/custom_issues_controller.rb (hunk 1 above)
Comment (MunteanuAndrei237) [reply app/controllers/custom_issues_controller.rb]: 1/ If I don't skip verify_authentication_token, I can't validate the AJAX request, even with the CSRF token
2/ I will move it

File: app/controllers/issue_view_columns_controller.rb [hunk 11]
```diff
@@ -1,15 +1,33 @@
 # frozen_string_literal: true
 
 class IssueViewColumnsController < ApplicationController
-  before_action :find_project_by_project_id
-  before_action :authorize
-  before_action :build_query_for_project
+  before_action :find_project_by_project_id, only: [:update]
```
Comment (mircea-negreanu) [inline app/controllers/issue_view_columns_controller.rb]: 1/ Actually, I want to make our entry special and let all the other to apply on all the rest.

E.g. if somebody add a new entry point I don't want to add to this line, I want that everything to be applied by default.

So put back the code that was deactivating the action for `:update_collapsed_info`

2/ Can you put then on a single line (deactivate `:find_project_by_project_id`, `:build_query_for_project` for `[:update_collapsed_ids]`) ?

File: app/controllers/issue_view_columns_controller.rb [hunk 12]
```diff
@@ -1,15 +1,33 @@
...
+  before_action :authorize, only: [:update]
+  before_action :build_query_for_project, only: [:update]
+  before_action :check_user_logged_in_and_has_rights, only: [:update_collapsed_ids]
```
Comment (mircea-negreanu) [inline app/controllers/issue_view_columns_controller.rb]: I don't think there is a need to use a separate function for this. You can check the code inside `update_collapsed_ids`.

As I can see it you call the `Issue.find` twice for the same record from db, which, in this case, is not ok.

File: app/controllers/issue_view_columns_controller.rb [hunk 13]
```diff
@@ -1,15 +1,33 @@
 # frozen_string_literal: true
 
 class IssueViewColumnsController < ApplicationController
-  before_action :find_project_by_project_id
-  before_action :authorize
-  before_action :build_query_for_project
+  before_action :find_project_by_project_id, only: [:update]
+  before_action :authorize, only: [:update]
```
Comment (mircea-negreanu) [inline app/controllers/issue_view_columns_controller.rb]: Please find a solution so that `:authorize` work. First, is this from redmine sources? Or is this only in this plugins / additionals?

Here https://medium.com/@codenode/how-to-use-remote-true-to-make-ajax-calls-in-rails-3ecbed40869b there seems to be an example on how to do it the rails way? Maybe there are others as well?

Just check what `:authorize` is doing and see what should be done.

File: app/helpers/issue_view_columns_issues_helper.rb [hunk 14]
```diff
@@ -6,49 +6,238 @@ def render_descendants_tree(issue)
...
+    end
+
+    # Retrieve sorting settings and determine if sorting by directory/file model is enabled
+    sort_dir_file_model = RedmineIssueViewColumns.setting(:sort_dir_file_model)
+    collapsed_ids = issue.collapsed_ids.to_s.split.map(&:to_i)
     field_values = +''
     s = table_start_for_relations columns_list
     manage_relations = User.current.allowed_to? :manage_subtasks, issue.project
-    # set data
-    issue_list(issue.descendants.visible.preload(:status, :priority, :tracker, :assigned_to).sort_by(&:lft)) do |child, level|
```
Comment (mircea-negreanu) [inline app/helpers/issue_view_columns_issues_helper.rb:14]: I'm wondering if this is translated into a query to database which sorts directly there by the `lft` field.
For more info on `lft` and `rgt` see https://github.com/collectiveidea/awesome_nested_set.

If it goes to the database, I'm wondering how feasible is to try and modify the `sort_by` to take our conditions into account. Can take a look and see if it feasible?

File: app/helpers/issue_view_columns_issues_helper.rb [hunk 15]
```diff
@@ -107,10 +296,10 @@ def issue_columns_with_closed_issues?
     return true if issue_scope_with_closed? issue_scope
 
     @issue_columns_with_closed_issues = if issue_scope == 'without_closed_by_default'
-                                          RedminePluginKit.true? params[:with_closed_issues]
```
Comment (mircea-negreanu) [inline app/helpers/issue_view_columns_issues_helper.rb]: Please fix the indentation.

File: assets/javascripts/issue_view_columns.js [hunk 16]
```diff
@@ -4,4 +4,6 @@ $(function() {
   $('#available_settings_issue_list_defaults_column_names option[value="subject"]').remove();
   $('#tab-content-issue_view_columns #available_c option[value="tracker"]').remove();
   $('#tab-content-issue_view_columns #available_c option[value="subject"]').remove();
+
```
Comment (mircea-negreanu) [inline assets/javascripts/issue_view_columns.js]: If there are no modifications, leave the file in the initial state.

File: db/migrate/003_add_collapsed_ids_to_issues.rb [hunk 17]
```diff
@@ -0,0 +1,5 @@
+class AddCollapsedIdsToIssues < ActiveRecord::Migration[6.0]
+  def change
+    add_column :issues, :collapsed_ids, :string, limit: 1000
```
Comment (mircea-negreanu) [inline db/migrate/003_add_collapsed_ids_to_issues.rb:3]: Where did you find out that `:text` is not supported on oracle?

According to what it is written here https://api.rubyonrails.org/v7.1.3.4/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-add_column should be generally balable. It is true that I didn't find code for Oracle on rails sources

File: app/views/settings/_issue_view_columns_settings.html.slim [hunk 18]
```diff
@@ -12,8 +12,38 @@ p
...
+                                   tag_name: 'settings[columns_min_width]'
+  em.info
+    = l :info_columns_min_width
+
+br
+
+p
+  = additionals_settings_textfield :columns_sorting,
+                                   value: @settings[:columns_sorting],
+                                   label: l(:label_columns_sorting),
```
Comment (mircea-negreanu) [inline app/views/settings/_issue_view_columns_settings.html.slim]: This is missing from translation files.

File: db/migrate/003_add_collapsed_ids_to_issues.rb (hunk 17 above)
Comment (MunteanuAndrei237) [reply db/migrate/003_add_collapsed_ids_to_issues.rb:3]: string e tradus in varchar si e mai universal, e mult mai safe

File: app/helpers/issue_view_columns_issues_helper.rb (hunk 14 above)
Comment (MunteanuAndrei237) [reply app/helpers/issue_view_columns_issues_helper.rb:14]: It's not because we still need to render the issues recursevely because the hidden attribute is passed down from the parent when you collapse. Also ,"sort by" takes only one arg and it's complicated making a custom function that takes both lft and our sorting criteria in consideration .Since lft doesn't hold information about the level we are on , we need to additionally check for both issues compared to have the same parent.

File: app/views/collapse_expand/_collapse_expand.js.erb [hunk 19]
```diff
@@ -0,0 +1,119 @@
+<script type="text/javascript">
+let collapsedIds = '<%= @issue.collapsed_ids.to_s %>';
+let issueId = '<%= @issue.id %>';
+
+function collapseExpand(element) {
+  // Helper function to handle the display logic
+  function toggleIssues(targetIssue, targetIssueIdntLevel, show) {
```
Comment (mircea-negreanu) [inline app/views/collapse_expand/_collapse_expand.js.erb]: You can simplify this code by using `JQuery` which is alreay included by redmine, so you just need to call the code.

In my opinion, I think you could use jquery to get all the issues that are after the index you received, and have an indent bigger the the element by using `filter`, as in https://stackoverflow.com/a/2613671

Just get the div for all the issues, then all the tr that have than info, and alter showing.

Should greatly simplify the code.

File: app/helpers/issue_view_columns_issues_helper.rb [hunk 20]
```diff
@@ -6,49 +6,238 @@ def render_descendants_tree(issue)
...
+  def comparison_lambda(sorting_criteria)
+    lambda do |a, b|
+      sorting_criteria.each do |criterion|
+        column_name = criterion[:column_name]
+        direction = criterion[:direction] == "ASC" ? 1 : -1
+
+        if column_name.start_with?("cf_")
+          # Handle custom fields
+          cf_id = column_name.sub(/^cf_/, "")
+          a_value = CustomValue.where(customized_id: a.id, customized_type: "Issue", custom_field_id: cf_id).first&.value
```
Comment (mircea-negreanu) [inline app/helpers/issue_view_columns_issues_helper.rb]: Can't you get the info from issue without searching in the database? I remember having a code that gets a value from issue base on a field (where field is either normal field or customfield) meaning you don't need to diferentaite here between `cf_` or some thing, it just uses redmine to return the value.

File: config/locales/de.yml [hunk 21]
```diff
@@ -15,3 +15,13 @@ de:
   info_issue_view_columns_project_settings: Durch Aktivierung des Projekt-Modul "Spalten in Ticketansicht" kann man individuell für das Projekt Spalten für Untertickets und Ticketbeziehungen festlegen (falls diese vom der systemweiten Einstellung abweichen sollen).
     Wenn die systemweite Einstellung verwendet werden soll, muss das Projekt Module deaktiviert werden.
   label_issue_view_column: Ticketspalte
+  label_sort_dir_file_model: Sortierung nach Ordner-/Dateimodell
```
Comment (mircea-negreanu) [inline config/locales/de.yml:18]: Since we do not know german, remove the translations for the new labels.

File: assets/javascripts/issue_view_columns.js [hunk 22]
```diff
@@ -4,4 +4,4 @@ $(function() {
   $('#available_settings_issue_list_defaults_column_names option[value="subject"]').remove();
   $('#tab-content-issue_view_columns #available_c option[value="tracker"]').remove();
   $('#tab-content-issue_view_columns #available_c option[value="subject"]').remove();
-});
+});
```
Comment (mircea-negreanu) [inline assets/javascripts/issue_view_columns.js:7]: I have seen this multiple times. I think the file does not have a new line at the end. This indicates a modification when there is no functional one. Can you revert this "modification"?

File: test/functional/issues_controller_test.rb [hunk 23]
```diff
@@ -19,7 +19,7 @@ class IssuesControllerTest < RedmineIssueViewColumns::ControllerTest
   def setup
     prepare_tests
 
-    @global_settings = { 'column_names' => %w[created_on updated_on] }
+    @global_settings = { "column_names" => %w[created_on updated_on]}
```
Comment (mircea-negreanu) [inline test/functional/issues_controller_test.rb:22]: You forgot to revert these types of modifications for this file.

File: test/functional/issues_controller_test.rb [hunk 24]
```diff
@@ -149,4 +149,126 @@ def test_show_without_closed_subtasks
       assert_select "tr#issue-#{closed_issue.id}", count: 0
     end
   end
+
+  def test_render_issue_tree_dir_file_model
```
Comment (mircea-negreanu) [inline test/functional/issues_controller_test.rb:153]: You said that you also fixed some of their original test, but I can't seem to find any modifications to that effect. 

Could we discuss about this?

File: app/helpers/issue_view_columns_issues_helper.rb [hunk 25]
```diff
@@ -6,49 +6,208 @@ def render_descendants_tree(issue)
...
+      field_values << rendered_issues.join("").html_safe
+    end
+  end
+
+  def sort_issues(issues, columns_list)
+    columns_sorting_setting = RedmineIssueViewColumns.setting(:columns_sorting)
+    return issues unless columns_sorting_setting.present?
+
+    # Build sorting criteria as an array of hashes with keys :column_name and :direction
+    sorting_criteria = columns_sorting_setting.split(",").map do |column_setting|
```
Comment (mircea-negreanu) [inline app/helpers/issue_view_columns_issues_helper.rb:144]: I think  there has been a misunderstandig.

The setting will contain the internationalization of the column in English (E.g. `Updated` and not `updated_on`).
I don't see a code to obtain the field info bazed on it's English name.

Please take a look at our plugin `get_list_of_fields` which returns a list of columns (`field info`) based on the english translation of the column.

This will be a list of objects, which you cand check if one of them is `CustomField` by using `field_info.is_a?(QueryCustomFieldColumn)`. I don't like the check `cf_` because they cand change it in the future.

File: test/functional/issues_controller_test.rb (hunk 24 above)
Comment (MunteanuAndrei237) [reply test/functional/issues_controller_test.rb:153]: It had 3 tests that failed, because of routing , I just added " only: [] " to my code : resources :issue_view_columns, only: [] do
    member do
      patch 'update_collapsed_ids'
    end
  end

**Comment by mircea-negreanu** [comment]
Code was ported to the `Improved GUI` branch.
